### PR TITLE
CIRC-2111 Create API wrapping settings CRUD

### PR DESCRIPTION
**Pull request by alexanderkurash**
[CIRC-2111](https://folio-org.atlassian.net/browse/CIRC-2111)
## Purpose
Create API wrapping settings CRUD (created in CIRCSTORE-509)

## Approach
Create a new interface with new CRUD endpoints

### CODE REVIEW

File: ramls/circulation-settings.raml [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            text/plain:
+              example: "Internal server error"
+    put:
+      is: [ validate ]
+      body:
+        application/json:
+          type: circulation-setting
+      responses:
+        204:
+          description: "Circulation settings have been saved."
```
Comment (OleksandrVidinieiev) [inline ramls/circulation-settings.raml:89]:
```suggestion
          description: "Circulation settings have been saved"
```

File: ramls/circulation-settings.raml [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        500:
+          description: "Internal server error"
+          body:
+            text/plain:
+              example: "Internal server error"
+    delete:
+      is: [ validate ]
```
Comment (OleksandrVidinieiev) [inline ramls/circulation-settings.raml:96]:
```suggestion
      is: [validate]
```

File: src/main/java/org/folio/circulation/domain/CirculationSetting.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  private final String id;
+
+  @Getter
+  private final String name;
+
+  @Getter
+  private final JsonObject value;
+
+  public static CirculationSetting from(JsonObject representation) {
+    if (getProperty(representation, ID_FIELD) == null ||
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/CirculationSetting.java:42]: In this method we are extracting each property twice. Should we extract them into variables?

File: src/main/java/org/folio/circulation/domain/CirculationSetting.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      getProperty(representation, NAME_FIELD) == null ||
+      getObjectProperty(representation, VALUE_FIELD) == null ||
+      !containsOnlyKnownFields(representation)) {
+
+      log.info("from:: Circulation setting JSON is invalid: {}", representation);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/CirculationSetting.java:47]: warn or error?

File: src/main/java/org/folio/circulation/infrastructure/storage/CirculationSettingsRepository.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    return FetchSingleRecord.<CirculationSetting>forRecord(RECORDS_PROPERTY_NAME)
+      .using(circulationSettingsStorageClient)
+      .mapTo(CirculationSetting::from)
+      .whenNotFound(failed(new RecordNotFoundFailure(RECORDS_PROPERTY_NAME, id)))
+      .fetch(id);
+  }
+
+  public CompletableFuture<Result<MultipleRecords<CirculationSetting>>> findBy(String query) {
+
+    return circulationSettingsStorageClient.getManyWithRawQueryStringParameters(query)
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/infrastructure/storage/CirculationSettingsRepository.java:43]:
```suggestion
  public CompletableFuture<Result<MultipleRecords<CirculationSetting>>> findBy(String query) {
    return circulationSettingsStorageClient.getManyWithRawQueryStringParameters(query)
```

File: src/main/java/org/folio/circulation/resources/CirculationSettingsResource.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  @Override
+  void create(RoutingContext routingContext) {
+    final var context = new WebContext(routingContext);
+    final var clients = Clients.create(context, client);
+    final var circulationSettingsRepository = new CirculationSettingsRepository(clients);
+
+    final var incomingRepresentation = routingContext.body().asJsonObject();
+    setRandomIdIfMissing(incomingRepresentation);
+    final var circulationSetting = CirculationSetting.from(incomingRepresentation);
+    log.debug("replace:: Creating circulation setting: {}", circulationSetting);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/CirculationSettingsResource.java:44]:
```suggestion
    log.debug("create:: Creating circulation setting: {}", () -> circulationSetting);
```

File: src/main/java/org/folio/circulation/resources/CirculationSettingsResource.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+  @Override
+  void replace(RoutingContext routingContext) {
+    final var context = new WebContext(routingContext);
+    final var clients = Clients.create(context, client);
+    final var circulationSettingsRepository = new CirculationSettingsRepository(clients);
+
+    final var incomingRepresentation = routingContext.body().asJsonObject();
+    final var circulationSetting = CirculationSetting.from(incomingRepresentation);
+    log.debug("replace:: Replacing circulation setting : {}", circulationSetting);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/CirculationSettingsResource.java:62]:
```suggestion
    log.debug("replace:: Replacing circulation setting : {}", () -> circulationSetting);
```

File: src/main/java/org/folio/circulation/resources/CirculationSettingsResource.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+      .thenApply(r -> r.map(JsonHttpResponse::created))
+      .thenAccept(context::writeResultToHttpResponse);
+  }
+
+  @Override
+  void get(RoutingContext routingContext) {
+    final var context = new WebContext(routingContext);
+    final var clients = Clients.create(context, client);
+    final var circulationSettingsRepository = new CirculationSettingsRepository(clients);
+
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/CirculationSettingsResource.java:77]: Add logging here and in the rest of methods below

File: src/main/java/org/folio/circulation/resources/CirculationSettingsResource.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    return r -> r.failWhen(id -> succeeded(!uuidIsValid(id)),
+      circulationSetting -> singleValidationError("Circulation setting ID is not a valid UUID",
+        "", ""));
+  }
+
+  private static boolean uuidIsValid(String providedId) {
+    try {
+      return providedId != null && providedId.equals(UUID.fromString(providedId).toString());
+    } catch(IllegalArgumentException e) {
+      log.debug("uuidIsValid:: Invalid UUID");
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/resources/CirculationSettingsResource.java:147]: warn or error?

Comment (alexanderkurash) [reply src/main/java/org/folio/circulation/resources/CirculationSettingsResource.java]: done

File: src/main/java/org/folio/circulation/domain/CirculationSetting.java (hunk 3 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/circulation/domain/CirculationSetting.java]: done

File: src/main/java/org/folio/circulation/domain/CirculationSetting.java (hunk 4 above)
Comment (alexanderkurash) [reply src/main/java/org/folio/circulation/domain/CirculationSetting.java]: done

**Review by OleksandrVidinieiev — APPROVED** [review]

**Review by roman-barannyk — APPROVED** [review]

**Pull request merged**